components/raftstore/src/store/fsm/peer.rs

```diff
@@ -753,157 +743,126 @@ where
-        if to_be_removed.is_empty()
-            && self.region().get_start_key() == region.get_start_key()
-            && self.region().get_end_key() == region.get_end_key()
-        {
-            // Nothing to be updated, return directly.
+    fn on_unsafe_recovery_demote_failed_voters(
+        &mut self,
+        shared_state: UnsafeRecoveryExecutePlanSharedState,
+        failed_voters: Vec<metapb::Peer>,
+    ) {
+        if self.fsm.peer.force_leader.is_none() {
+            error!(
+                "Unsafe recovery, demoting failed voters failed, since this peer is not forced leader";
+                "region" => ?self.region(),
+            );
             return;
         }
-        info!(
-            "updating the reigon for unsafe recover, original: {:?}, target: {:?}",
-            self.region(),
-            region
-        );
-        if self.fsm.peer.has_valid_leader() {
-            panic!("region update for unsafe recover should only occur in leaderless reigons");
-        }
-        if self.fsm.peer.raft_group.store().applied_index()
-            != self.fsm.peer.raft_group.store().commit_index()
-        {
+
+        if self.fsm.peer.unsafe_recovery_state.is_some() {
             warn!(
-                "cannot proceed region update for unsafe recover, applied index is not equal to commit index"
+                "Unsafe recovery, demote failed voters has already been initiated";
+                "region" => ?self.region(),
             );
             return;
         }
 
-        let region_state_key = keys::region_state_key(region.get_id());
-        let original_region_state = match self
-            .ctx
-            .engines
-            .kv
-            .get_msg_cf::<RegionLocalState>(CF_RAFT, &region_state_key)
-        {
-            Ok(Some(region_state)) => region_state,
-            Ok(None) => {
-                panic!("Can't find RegionLocalState while updating {:?}", region);
-            }
-            Err(e) => {
-                panic!(
-                    "Fail to look up RegionLocalState while updating {:?} err {:?}",
-                    region, e
-                );
-            }
-        };
-        let mut kv_wb = self.ctx.engines.kv.write_batch();
-        write_peer_state(&mut kv_wb, &region, PeerState::Normal, None).unwrap_or_else(|e| {
-            panic!(
-                "fails to write RegionLocalState {:?} into write brach, err {:?}",
-                region, e
-            )
-        });
-        let mut write_opts = WriteOptions::new();
-        write_opts.set_sync(true);
-        if let Err(e) = kv_wb.write_opt(&write_opts) {
-            panic!("fail to update RegionLocalstate {:?} err {:?}", region, e);
-        }
-
-        {
-            let mut meta = self.ctx.store_meta.lock().unwrap();
-            meta.set_region(
-                &self.ctx.coprocessor_host,
-                region.clone(),
-                &mut self.fsm.peer,
+        if self.fsm.peer.in_joint_state() {
```

Comment:
How about introducing PreDecomeFailedVoters for unsafe_recovery_state, then some code can be reused